src/nxt_cgroup.c

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) Andrew Clayton
+ * Copyright (C) F5, Inc.
+ */
+
+#include <nxt_main.h>
+#include <nxt_process.h>
+#include <nxt_fs.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <limits.h>
+
+
+static nxt_int_t nxt_cgroup_create(const nxt_task_t *task,
+    nxt_process_t *process);
+
+
+static nxt_int_t
+nxt_cgroup_create(const nxt_task_t *task, nxt_process_t *process)
+{
+    char  cgpath[PATH_MAX];
+
+    if (task->thread->runtime->type != NXT_PROCESS_MAIN
+        || nxt_process_type(process) != NXT_PROCESS_PROTOTYPE
+        || process->isolation.cgroup.name == NULL)
+    {
+        return NXT_OK;
+    }
+
+    snprintf(cgpath, sizeof(cgpath), NXT_CGROUP_ROOT "/%s",
+             process->isolation.cgroup.name);
+    nxt_fs_mkdir_all((const u_char *)cgpath, 0777);
+
+    return NXT_OK;
+}
+
+
+nxt_int_t
+nxt_cgroup_proc_add(const nxt_task_t *task, nxt_process_t *process)
+{
+    FILE  *fp;
+    char  cgprocs[PATH_MAX];
+
+    nxt_cgroup_create(task, process);
+
+    if (nxt_process_type(process) != NXT_PROCESS_PROTOTYPE
+        || process->isolation.cgroup.name == NULL)
+    {
+        return NXT_OK;
+    }
+
+    snprintf(cgprocs, sizeof(cgprocs), NXT_CGROUP_ROOT "/%s/cgroup.procs",
+             process->isolation.cgroup.name);
+    fp = fopen(cgprocs, "we");
+    if (fp == NULL) {
+        return NXT_ERROR;
+    }
+
+    fprintf(fp, "%d\n", process->pid);
+    fclose(fp);
+
+    return NXT_OK;
+}
+
+
+void
+nxt_cgroup_cleanup(const nxt_process_t *process)
+{
+    char  *ptr;
+    char  cgpath[PATH_MAX];
+
+    snprintf(cgpath, sizeof(cgpath), NXT_CGROUP_ROOT "/%s",
+             process->isolation.cgroup.name);
+    while (*cgpath != '\0' && strcmp(NXT_CGROUP_ROOT, cgpath) != 0) {
+        rmdir(cgpath);
```

Comment:
The error should not be silently ignored here because if the cgroup path was successfully created then it should be cleaned up or log an error for troubleshooting.
I know there are some decisions yet to be made regarding the error handling, but maybe having a `process->has_own_cgroup` set to `1` when it did exist (creation succeed) so you can guard the `rmdir()` and then log the error.

```
void
nxt_cgroup_cleanup(const nxt_process_t *process)
{
    char  *ptr;
    char  cgpath[PATH_MAX];

    if (!process->has_own_cgroup) {
        return NXT_OK;
    }

    ...
}
```

Reply:
Possibly. Though you'd need to handle the case where *some* of the path was created...

I was currently taking the approach of trying to clean things up, but no big worry if not as it's transitory.